### PR TITLE
feat: implement ludwig nlu helper extension

### PR DESCRIPTION
**Pull request by MohamedAliBouhaouala**

### CODE REVIEW

File: i18n/en/label.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "endpoint": "Ludwig NLU API",
+    "token": "API Token",
```
Comment (marrouchi) [inline i18n/en/label.json:3]: We don't have any API token (currently) for Ludwig, right ? If that is the case then we can remove this setting.

File: index.helper.ts [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+   * @returns The formatted Rasa NLU training dataset.
+   */
+  async format(
+    samples: NlpSampleFull[],
+    entities: NlpEntityFull[],
+  ): Promise<RasaNlu.Dataset> {
+    const entityMap = NlpEntity.getEntityMap(entities);
+    const valueMap = NlpValue.getValueMap(
+      NlpValue.getValuesFromEntities(entities),
+    );
```
Comment (marrouchi) [inline index.helper.ts]: We need to adapt the format for Ludwig

File: index.helper.ts [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    // Train samples
+    return await this.httpService.axiosRef.post(
+      buildURL(settings.endpoint, `/train`),
+      nluData,
+      {
+        params: {
+          token: settings.token,
+        },
+      },
+    );
```
Comment (marrouchi) [inline index.helper.ts]: Throw new Error('Method not implemented yet!')

Same for evaluate()

File: index.helper.ts [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      nluTestData,
+      {
+        params: {
+          token: settings.token,
+        },
+      },
+    );
+  }
+
+  async process_incoming_nlu_payload(
```
Comment (marrouchi) [inline index.helper.ts]: JS is in camelCase ;)

File: index.helper.ts [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+                : e.confidence;
+            return e;
+          })
+          .filter((e) => e.confidence >= minConfidence);
+        // Get past threshold and the highest confidence for the same entity
+        // .filter((e, idx, self) => {
+        //   const sameEntities = self.filter((s) => s.entity === e.entity);
+        //   const max = Math.max.apply(Math, sameEntities.map((e) => { return e.confidence; }));
+        //   return e.confidence === max;
+        // });
```
Comment (marrouchi) [inline index.helper.ts]:
```suggestion

```

File: index.helper.ts [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+            entity: trait,
+            value: (nlp as any)[trait].name,
+            confidence: (nlp as any)[trait].confidence,
+          });
+        }
+      });
+      return guess;
+    } catch (e) {
+      this.logger.error(
+        'Core NLU Helper : Unable to parse nlp result to extract best guess!',
```
Comment (marrouchi) [inline index.helper.ts]:
```suggestion
        'Ludwig NLU Helper : Unable to parse nlp result to extract best guess!',
```

File: package.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+    "name": "hexabot-helper-ludwig-nlu",
```
Comment (marrouchi) [inline package.json]:
```suggestion
    "name": "hexabot-helper-ludwig",
```

File: types.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2024 Hexastack. All rights reserved.
+ *
+ * Licensed under the GNU Affero General Public License v3.0 (AGPLv3) with the following additional terms:
+ * 1. The name "Hexabot" is a trademark of Hexastack. You may not use this name in derivative works without express written permission.
+ * 2. All derivative works must include clear attribution to the original creator and software, Hexastack and Hexabot, in a prominent location (e.g., in the software's "About" section, documentation, and README file).
+ */
+
+export namespace RasaNlu {
```
Comment (marrouchi) [inline types.ts]: Please update typing accordingly.

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: i18n/en/label.json (hunk 1 above)
Comment (MohamedAliBouhaouala) [reply i18n/en/label.json:3]: It's just a token for the Hexabot API to interact with the Ludwig NLU engine

File: index.helper.ts [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    const valueMap = NlpValue.getValueMap(
+      NlpValue.getValuesFromEntities(entities),
+    );
+
+    const dataset: LudwigNlu.LudwigNluDataSample[] = samples
+      .filter((s) => s.entities.length > 0)
+      .map((s) => {
+        const intent = s.entities.find(
+          (e) => entityMap[e.entity].name === 'intent',
+        );
```
Comment (marrouchi) [inline index.helper.ts]:
```suggestion
        const traitEntities = s.entities.find(
          (e) => entityMap[e.entity].lookups.includes('trait'),
        );
```

File: index.helper.ts [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+  ): Promise<LudwigNlu.NluProcessedResultType> {
+    const words = givenText.split(' ');
+
+    const intentValue = nlp.intent.predictions.intent_predictions;
+    const intentKey = `intent_probabilities_${intentValue}`; // e.g., "intent_probabilities_greeting"
+    const intentConfidence = nlp.intent.predictions[intentKey];
+    const formattedIntentPayload = {
+      name: intentValue,
+      confidence: intentConfidence,
+    };
```
Comment (marrouchi) [inline index.helper.ts]: .We need to parse the results depending on the trait entities we have (we may not only have intent)

**Review by marrouchi — CHANGES_REQUESTED** [review]

**Review by marrouchi — APPROVED** [review]

**Pull request merged**